modules/data.remote/R/download.thredds.R

```diff
@@ -20,84 +30,124 @@
-  suppressWarnings(qry_results <- DBI::dbFetch(qry_results))
-  site_info <- list(site_id=qry_results$id, site_name=qry_results$sitename, lat=qry_results$lat,
-                    lon=qry_results$lon, time_zone=qry_results$time_zone)
+  #remove files with no dates and get list of dates available.
+  index_dates = regexpr(pattern = "[0-9]{8}", files)
+  files = files[-(which(index_dates < 0))]
+  index_dates = index_dates[which(index_dates > 0)]
   
-  mylat = site_info$lat
-  mylon = site_info$lon
+  # get list of files that fall within the specific date range user asks for (Ymd, not Y)
+  dates_avail = as.Date(substr(files, index_dates, index_dates+7), "%Y%m%d")
+  date_range = seq(dates[1], dates[2], by = "day")
+  get_dates =  date_range[which(date_range %in% dates_avail)]
   
-  # site specific URL for dataset --> these will be made to work for all THREDDS datasets in the future, but for now, just testing with
-  # this one dataset. This specific dataset only has 1 year (2005), so no temporal looping for now.
-  obs_file = "https://thredds.daac.ornl.gov/thredds/dodsC/ornldaac/1221/agb_5k.nc4"
-  obs_err = "https://thredds.daac.ornl.gov/thredds/dodsC/ornldaac/1221/agb_SE_5k.nc4"
-  files = c(obs_file, obs_err)
+  # only keep files that are within the true yyyymmdd date range user requested
+  files = files[foreach(i = seq_along(get_dates), .combine = c) %do% grep(files, pattern = format(get_dates[i], '%Y%m%d'))]
+  filenames = basename(files)
   
-  # function to extract ncdf data from lat and lon values for value + SE URLs
-  get_data = function(i)
+  # user must supply data_URL or the netcdf files cannot be downloaded through thredds. if user has supplied no data_url, the job will fail
+  # supply a warning
+  if (!(is.null(data_url)))
   {
-    data = ncdf4::nc_open(files[1])
-    agb_lats = ncdf4::ncvar_get(data, "latitude")
-    agb_lons = ncdf4::ncvar_get(data, "longitude")
-    
-    agb_x = which(abs(agb_lons- mylon[i]) == min(abs(agb_lons - mylon[i])))
-    agb_y = which(abs(agb_lats- mylat[i]) == min(abs(agb_lats - mylat[i])))
-   
-    start = c(agb_x, agb_y)
-    count = c(1,1)
-    d  = ncdf4::ncvar_get(ncdf4::nc_open(files[1]), "abvgrndbiomass", start=start, count = count)
-    if (is.na(d)) d <- NA
-    sd = ncdf4::ncvar_get(ncdf4::nc_open(files[2]), "agbSE", start=start, count = count)
-    if (is.na(sd)) sd <- NA
-    date = "2005"
-    site = site_ID[i]
-    output = as.data.frame(cbind(d, sd, date, site))
-    names(output) = c("value", "sd", "date", "siteID")
+    #https://www.ncei.noaa.gov/thredds/dodsC/cdr/lai/files/1981/AVHRR-Land_v005_AVH15C1_NOAA-07_19810624_c20181025194251.nc.html
+    # this is what a link looks like to download threeds data.
+    urls = sort(paste(data_url, substr(dates_avail, 1, 4), filenames, sep = "/"))
     
-    # option to save output dataset to directory for user.
-    if (!(is.null(outdir)))
+    extract_nc = function(site_info, url, run_parallel)
```

Comment:
I'll also add that there's already an extract.nc function in data.atmosphere, which is going to be very confusing. Furthermore, after we clean up the function naming (i.e. finally switch . to _ ) it could easily lead to genuine conflicts. One option is to rename this function, since it's the new one, and because the args makes it clear that this function isn't just extracting it's also downloading. Alternatively, you could take a look at the other function and think about whether you could generalize it (without breaking it's current I?O and behaviors) to serve both cases.

Either way, the function definitely needs Roxygen documentation, especially since site_info is a list of stuff

Reply:
Good point, @mdietze -- renaming or generalizing seems wise. But do note that this is a helper that isn't exported and probably doesn't need to be, which (1) mitigates the name conflict issues (though not the potential for confusion) and (2) personally I think it makes Roxygen docs less important -- to me it's clear in context that site_info will be the same thing here as it is in download_thredds_data.